File: q21.py
"""
21. Merge Two Sorted Lists

Merge two sorted linked lists and return it as a sorted list.
The list should be made by splicing together the nodes of the first two lists.

https://leetcode.com/problems/merge-two-sorted-lists/
"""


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def mergeTwoLists(self, l1: ListNode, l2: ListNode) -> ListNode:
        dummy_head = ListNode(0, None)
        head = dummy_head
        while l1 is not None or l2 is not None:
            if l1 is None:
                # add l2
                head.next = l2
                l2 = l2.next
            elif l2 is None:
                # add l1
                head.next = l1
                l1 = l1.next
            elif l1.val < l2.val:
                # add l1
                head.next = l1
                l1 = l1.next
            else:
                # add l2
                head.next = l2
                l2 = l2.next
            head = head.next
        return dummy_head.next

File: q21.cpp
/**
 * 21. Merge Two Sorted Lists
 * https://leetcode.com/problems/merge-two-sorted-lists
 */

#include "my_util.hpp"
#include "my_list_node.hpp"

using namespace std;

using ListNode = ListNode_generic<int>;

class Solution {
public:
    ListNode *mergeTwoLists(ListNode *list1, ListNode *list2) {
        // must directly merge the two list, without copying

        ListNode dummyHead{-1, nullptr}, *head = &dummyHead;
        while (list1 != nullptr && list2 != nullptr) {
            if (list1->val < list2->val) {
                head->next = list1;
                head = list1;
                list1 = list1->next;
            } else {
                head->next = list2;
                head = list2;
                list2 = list2->next;
            }
        }

        if (list1 != nullptr) {
            head->next = list1;
        }
        if (list2 != nullptr) {
            head->next = list2;
        }

        print_varl(sep, dummyHead);

        return dummyHead.next;
    }

};

int main() {
    make_list_node(int, i, 1, 2, 4);
    make_list_node(int, j, 1, 3, 4);
    print_v(i0, j0);
    print_line(Solution().mergeTwoLists(&i0, &j0));
}
